pkg/resources/external_volume.go

```diff
@@ -0,0 +1,717 @@
+		} else {
+			// no way to unset allow writes - set to false as a default
+			set.WithAllowWrites(false)
+		}
+	}
+
+	if (*set != sdk.AlterExternalVolumeSetRequest{}) {
+		if err := client.ExternalVolumes.Alter(ctx, sdk.NewAlterExternalVolumeRequest(id).WithSet(*set)); err != nil {
+			return diag.FromErr(err)
+		}
+	}
+
+	if d.HasChange("storage_location") {
+		old, new := d.GetChange("storage_location")
+		oldLocations, err := extractStorageLocations(old, true)
+		if err != nil {
+			return diag.FromErr(err)
+		}
+
+		newLocations, err := extractStorageLocations(new, false)
+		if err != nil {
+			return diag.FromErr(err)
+		}
+
+		// Storage locations can only be added to the tail of the list, but can be
+		// removed at any position. Given this limitation, to keep the configuration order
+		// matching that on Snowflake the list needs to be partially recreated. For example, if a location
+		// is added in the configuration at index 5 in the list, all existing storage locations from index 5
+		// need to be removed, then the new location can be added, and then the removed locations
+		// can be added back. The storage locations lower than index 5 don't need to be modified.
+		// The removal process could be done without the above recreation, but it handles this case
+		// too so it's used for both actions.
+		longestCommonPrefix, err := LongestCommonPrefix(newLocations, oldLocations)
+		if err != nil {
+			return diag.FromErr(err)
+		}
+
+		var removedLocations []sdk.ExternalVolumeStorageLocation
+		var addedLocations []sdk.ExternalVolumeStorageLocation
+		if longestCommonPrefix == -1 {
+			removedLocations = oldLocations
+			addedLocations = newLocations
+		} else {
+			// Could +1 on the prefix here as the lists until and including this index
+			// are identical, would need to add some more checks for list length to avoid
+			// an array index out of bounds error
+			removedLocations = oldLocations[longestCommonPrefix:]
+			addedLocations = newLocations[longestCommonPrefix:]
+		}
+
+		if len(removedLocations) == len(oldLocations) {
+			// Create a temporary storage location, which is a copy of a storage location currently existing
+			// except with a different name. This is done to avoid recreating the external volume, which
+			// would otherwise be necessary as a minimum of 1 storage location per external volume is required.
+			// The alternative solution of adding volumes before removing them isn't possible as
+			// name must be unique for storage locations
+			if len(removedLocations) > 1 {
+				// To ensure the name of the temporary storage location is unique,
+				// first remove all but 1 storage locations from the list
+				for _, removedStorageLocation := range removedLocations[1:] {
+					err = removeStorageLocation(removedStorageLocation, client, ctx, id)
+					if err != nil {
+						return diag.FromErr(err)
+					}
+				}
+			}
+
+			temp_storage_location, err := CopyStorageLocationWithTempName(removedLocations[0])
```

Comment:
It shouldn’t happen often but if deleting the temp resources fails in one apply, adding them on the next apply would fail due to duplicate names. If we want to avoid this, we could first run a check to see if the temp storage location already exists and run a remove operation.